include/envoy/network/io_handle.h

```diff
@@ -7,27 +7,93 @@
 namespace Envoy {
 namespace Network {
 
+namespace Address {
+class Instance;
+} // namespace Address
+
+/**
+ * Basic type for return result which has a return code and error code defined
+ * according to different implementation.
+ */
+template <typename T> struct IoHandleCallResult {
+  virtual ~IoHandleCallResult() {}
+
+  T rc_;
+  int errno_;
+};
+
+typedef IoHandleCallResult<int> IoHandleCallIntResult;
+typedef IoHandleCallResult<ssize_t> IoHandleCallSizeResult;
+
 /**
  * IoHandle: an abstract interface for all I/O operations
  */
 class IoHandle {
 public:
-  IoHandle() {}
-
   virtual ~IoHandle() {}
 
   /**
    * Return data associated with IoHandle.
    *
    * TODO(sbelair2) remove fd() method
+   * We probably still need some method similar to this one for
+   * evconnlistener_new(). Or We can move it to IoSocketHandle and down cast the
+   * IoHandle to IoSocketHandle wherever needed.
    */
   virtual int fd() const PURE;
 
   /**
    * Clean up IoHandle resources
    */
-  virtual void close() PURE;
+  virtual IoHandleCallIntResult close() PURE;
+
+  virtual bool isClosed() PURE;
+
+  virtual IoHandleCallSizeResult readv(const iovec* iovec, int num_iovec) PURE;
+
+  virtual IoHandleCallSizeResult writev(const iovec* iovec, int num_iovec) PURE;
+
+  virtual IoHandleCallIntResult bind(const Network::Address::Instance& address) PURE;
+
+  virtual IoHandleCallIntResult connect(const Network::Address::Instance& server_address) PURE;
+
+  /**
+   * Wrap setsockopt()
+   */
+  virtual IoHandleCallIntResult setIoHandleOption(int level, int optname, const void* optval,
+                                                  socklen_t optlen) PURE;
+  /**
+   * Wrap getsockopt()
+   */
+  virtual IoHandleCallIntResult getIoHandleOption(int level, int optname, void* optval,
+                                                  socklen_t* optlen) PURE;
+
+  /**
+   * Wrap getsockname()
+   */
+  virtual IoHandleCallIntResult getIoHandleName(const Network::Address::Instance& address) PURE;
+
+  virtual IoHandleCallIntResult getPeerName(const Network::Address::Instance& address) PURE;
+
+  /**
+   * Wrap fcntl(fd_, F_SETFL...)
+   */
+  virtual IoHandleCallIntResult setIoHandleFlag(int flag) PURE;
+  /**
+   * Wrap fcntl(fd_, F_GETFL...)
+   */
+  virtual IoHandleCallIntResult getIoHandleFlag() PURE;
+
+  virtual IoHandleCallIntResult listen(int backlog) PURE;
+
+  /**
+   * Wrap dup()
+   */
+  virtual std::unique_ptr<IoHandle> dup() PURE;
+
+  virtual IoHandleCallIntResult shutdown(int how) PURE;
 };
```

Comment:
It seems to me that there is probably no point in defining the template class IoHandleCallResult if all the virtual methods in the base class are all IoHandleCallResult<int> or IoHandleCallResult<ssize_t>. I think we need to examine if there are any return values that can be other than integer types. If there are other interfaces that need this (I'm not saying that VPP is one of them), then won't these virtual methods preclude them?

Reply:
@danzh2010 I do want to say that I like the direction where this is going.

Reply:
Are you concerned about some implementations of same method whose return a type is different from what is declared here? For such method, I would say same purpose can be achieved by adding another new method to the interface. 

If IoHandleCallResult<int> and IoHandleCallResult<ssize_t> are the only two return types we need, what would you like the IoHandleCallResult  to be without template?

Reply:
@danzh2010 Sorry, forgot about this. I'm only thinking that we should not preclude further options, but maybe that's just too much caution? Right now I cannot think of other return types we might need, but that doesn't mean that such a need won't come up later.

Reply:
Now we have IoHandleCallResult<std::bitset<2>>